### PR TITLE
several bugfixes and tweaks

### LINKED ISSUE #243: Cannot get the correct Tile16 number of the last Tile16 in layers from the script interface
use interface.log(interface.GetCurRoomTile16(1, RoomWidth, RoomHeight)); in the js script always get 0xFFFF as a result.

### LINKED ISSUE #245: Changing Layer 0 Mapping Type to Tile8x8 related value will crash the editor
- Open a ROM
- Load Level: Sapphire Passage - Third Level (Fiery Cavern)
- Switch to Room 2
- Open Room Config
- Change Layer 0 Mapping Type to "0"(Disabled) from "22"(Tile8x8 & autoscroll)
- Press OK (close Room Config)
- Open Room Config
- Change Layer 0 Mapping Type to "22"(Tile8x8 & autoscroll) from "0"(Disabled)
- Press OK (close Room Config)
- The editor crashes

### LINKED ISSUE #250: layer 0 alpha blending cannot be rendered correctly in the editor on transparent pixels
according to the gbatek documentations:
the top-most non-transparent pixel must be selected as 1st Target, and the next-lower non-transparent pixel must be selected as 2nd Target, if so - and only if so, then color intensities of 1st and 2nd Target are mixed together by using the parameters in BLDALPHA register.

so we should disable alpha blending for pixels if pixels in either layers are transparent.

### PR DESCRIPTION
**Pull request by shinespeciall**
Fixed #243 cannot get the correct Tile16 number of the last Tile16 in layers from the script interface.
Fixed #245 Changing Layer 0 Mapping Type to Tile8x8 related value will crash the editor
Fixed #250 layer 0 alpha blending cannot be rendered correctly in the editor on transparent pixels
Changes in Door Config Dialog: Door type and Entity names text
Changes in Room Config Dialog: Raster type and water text
Rewrote animated tile loading logic comment in Tileset.cpp, can be used to make animated tile editor and do relative tweaks on the code
Tweaks on script APIs

### CODE REVIEW

File: Dialog/DoorConfigDialog.h [hunk 1]
```diff
@@ -221,26 +221,26 @@ private slots:
...
         "0x77 Dice",
-        "0x78 Toy Car 10 points",
+        "0x78 Toy Car",
         "0x79 Onomī",
         "0x7A Cat in boss corridor",
         "0x7B PET Bottle",
-        "0x7C Pearl Bird",
-        "0x7D Boss: Golden Diva",
+        "0x7C mouja",
+        "0x7D Boss: Golden Diva, yuka",
```
Comment (Goldensunboy) [inline Dialog/DoorConfigDialog.h:243]: Yōki

File: Dialog/RoomConfigDialog.cpp [hunk 2]
```diff
@@ -234,6 +257,11 @@ void RoomConfigDialog::on_ComboBox_TilesetID_currentIndexChanged(int index)
             ui->ComboBox_Layer0Picker->addItem(
                 QString::number(WL4Constants::ToxicLandfillDustyLayer0Ptr, 16).toUpper());
         }
+        else if (ui->ComboBox_TilesetID->currentIndex() == 0x45)
```
Comment (Goldensunboy) [inline Dialog/RoomConfigDialog.cpp:260]: Let's make this a constant (and if it is used somewhere else, also change that usage)

File: LevelComponents/Room.cpp [hunk 3]
```diff
@@ -352,17 +352,25 @@ namespace LevelComponents
...
-                                                 (Layer0ColorBlendCoefficient_EVB * PXB.red()) / 16,
-                                             255);
-                                int G = qMin((Layer0ColorBlendCoefficient_EVA * PXA.green()) / 16 +
-                                                 (Layer0ColorBlendCoefficient_EVB * PXB.green()) / 16,
-                                             255);
-                                int B = qMin((Layer0ColorBlendCoefficient_EVA * PXA.blue()) / 16 +
-                                                 (Layer0ColorBlendCoefficient_EVB * PXB.blue()) / 16,
-                                             255);
-                                imageA.setPixel(k, j, QColor(R, G, B).rgb());
+                                if(imageA.pixelColor(k, j).alpha() == 0xFF && imageB.pixelColor(k, j).alpha() == 0xFF) // current pixels not transparent
```
Comment (Goldensunboy) [inline LevelComponents/Room.cpp:355]: Let's extract this to a function (and the 2 other instances of the same logic)

File: ScriptInterface.cpp [hunk 4]
```diff
@@ -8,28 +8,42 @@ ScriptInterface::ScriptInterface(QObject *parent) : QObject(parent)
...
-    if(room->GetLayer(layerID)->GetMappingType() != LevelComponents::LayerMap16)
+    LevelComponents::Layer *layer = singleton->GetCurrentRoom()->GetLayer(layerID);
+    if(layer->GetMappingType() != LevelComponents::LayerMap16)
         return -1;
-    int width = static_cast<int>(room->GetWidth());
-    int height = static_cast<int>(room->GetHeight());
+    int width = static_cast<int>(layer->GetLayerWidth());
+    int height = static_cast<int>(layer->GetLayerHeight());
+    if(x >= width || y >= height) {
+        log(QString("position out of range!\n"));
```
Comment (Goldensunboy) [inline ScriptInterface.cpp:29]: Let's make this more descriptive. Such as, "ScriptInterface::GetCurRoomTile16(): Tile position (x, y) out of bounds on layer L (dimensions: _x, _y)" etc and of course replace those letter with values

Comment (Goldensunboy) [reply ScriptInterface.cpp:29]: And also do this in ScriptInterface::GetCurRoomTile8()

File: WL4EditorWindow.cpp [hunk 5]
```diff
@@ -563,6 +544,10 @@ void WL4EditorWindow::RoomConfigReset(DialogParams::RoomConfigParams *currentroo
     {
         currentRoom->GetLayer(0)->CreateNewLayer_type0x10(nextroomconfig->Layer0Width, nextroomconfig->Layer0Height);
     }
+    if ((currentroomconfig->Layer0MappingTypeParam > 0x10) && (nextroomconfig->Layer0MappingTypeParam < 0x10))
```
Comment (Goldensunboy) [inline WL4EditorWindow.cpp:547]: In this whole function, can you please add a comment before each if-else block to tell what it does (do not need before "else", just before the "if")

File: Dialog/DoorConfigDialog.h (hunk 1 above)
Comment (shinespeciall) [reply Dialog/DoorConfigDialog.h:243]: Done.

File: WL4EditorWindow.cpp (hunk 5 above)
Comment (shinespeciall) [reply WL4EditorWindow.cpp:547]: Done.

File: ScriptInterface.cpp (hunk 4 above)
Comment (shinespeciall) [reply ScriptInterface.cpp:29]: Done.

File: Dialog/RoomConfigDialog.cpp (hunk 2 above)
Comment (shinespeciall) [reply Dialog/RoomConfigDialog.cpp:260]: Done weirdly.

File: LevelComponents/Room.cpp (hunk 3 above)
Comment (shinespeciall) [reply LevelComponents/Room.cpp:355]: Done.

**Pull request merged**